docs: add release docs (#1087)
VeryGoodOpenSource/dart_frog: CONTRIBUTING.md

First off, thanks for taking the time to contribute! 🎉👍

These are mostly guidelines, not rules. Use your best judgment, and feel free to propose changes to this document in a pull request.

This project is opinionated and follows patterns and practices used by the team at [Very Good Ventures][very_good_ventures_link].

## Understanding the Dart Frog repository

This is a mono repo. This means that this repository includes more than one individual project. In fact, the Dart Frog repository includes all the packages, example apps, CLIs, and IDE integration plugins that have a role in the Dart Frog developer experience. 

The contents of the repo are divided into the following repositories: 

- `tool/`: contains internal operation scripts

> - `assets/`: images to embed into readmes

Reviewer: **Nit:**
```suggestion
- `assets/`: images to embed into READMEs
```

work, tests, or other changes before your pull request can be ultimately
accepted.

> # For maintainers

Reviewer: ```suggestion
# Developing for Dart Frog
```

In Very Good CLI and other contributing files we name the section "Developing for X". It was inspired by Flutter's CONTRIBTUING, see [this](https://github.com/flutter/flutter/blob/master/CONTRIBUTING.md#developing-for-flutter).

> ## Setting up environment

Reviewer: ```suggestion
## Setting up your local development environment
```

Similar to the previous comment, other contributing files we have use the "Setting up your local development environment" as a header. It was also inspired by Flutter's CONTRIBUTING headers, see [this](https://github.com/flutter/flutter/wiki/Setting-up-the-Engine-development-environment) and [this](https://github.com/flutter/flutter/wiki/Setting-up-the-Framework-development-environment).

Prerequisites:

> - Dart SDK installed in the path, [standalone][dart_standalone_link], or included in the [Flutter SDK][dart_on_flutter_link].

Reviewer: We should inform what version of the Dart SDK the user should install (ideally forwarding them to the document where this is specified to avoid redundancy), in other CONTRIBUTING files we use a text like the one below:

```suggestion
- Install a valid [Dart SDK](https://dart.dev/get-dart) in your local environment, it should be compatible with the latest version of [Dart Frog CLI](https://github.com/VeryGoodOpenSource/dart_frog/blob/main/packages/dart_frog_cli/pubspec.yaml). If you have Flutter installed, you likely have a valid Dart SDK version already installed.
```

- [Mason CLI][mason_install_link] (to run and test the `bricks`);

> - [Node.js][node_js_dowload_link] (For the VS Code extension and the docs site);

Reviewer: I prefer forwarding the users to read the [VS Code's contributing file](https://github.com/VeryGoodOpenSource/dart_frog/blob/68b290698068e5a2446f1cea7fb68b8921861f07/extensions/vscode/CONTRIBUTING.md#setting-up-your-local-development-environment). Where it is documented what version of Node.js will they need and any other dependencies.

I'm not sure if the same Node.js version is compatible with the docs site to.

Author: Sure, how should we word that?

Reviewer: ```suggestion
- [Node.js][node_js_dowload_link], for working with the VS Code extension or the documentation website. Please refer to their CONTRIBUTING files for further installation requirements.
- [Yarn](https://classic.yarnpkg.com/lang/en/docs/install/), for working with the documentation website.
```

- Capability to run shell scripts (for the scripts under `tool/`).

> ## Understanding the `packages/` contents:

Reviewer: I think we could avoid this section here and moving any information to the package specific CONTRIBUTING, wdyt?

We're already referencing to them:
> "Some of the included projects have more specific instructions on contribution. In these cases, the project root may include a `CONTRIBUTING.md` file with such instructions. "

Author: I think the relationship between these packages is fundamental to understanding when/how to update them. The individual maintenance info is relevant too. Will add that paragraph.

### `dart_frog_gen`

This is the internal package used by the Dart Frog tooling to interpret the file disposition and from it construct a Dart Frog server. 

> :warning: **Warning**: this package is a dependency on the bricks bundled into the CLI. This means that any changes that break the bricks should be released with a major version, otherwise dart frog users may be blocked from performing tasks such as `dev`, `build`, and `new`.

### `dart_frog_cli`

A Dart command line interface package that serves as the main tool for Dart Frog. It includes bundled versions of the bricks under `bricks/`. To sync the source code of the bricks with new bundles, run `tool/generate_bundles.sh`.

### Companion packages

The other items under `packages/` are companion packages in which dart_frog users may include on their project for specific server-side capabilities, such as auth (`dart_frog_auth`) and WebSockets (`dart_frog_web_socket`)

## Releasing versions of packages

> Before starting the release process of an individual package, first check:
> 
> - If your local `main` branch is up to date;
> - If the pipeline is green for the given package;
> - How this release will possibly affect other packages. Make sure to open follow-up PRs updating this package usage in other Dart Frog packages that depend on it, after the release.
> 
> Checkout on `main` and run the script under `./tool/release_ready.sh` in the package root repository with the desired new version. 
> 
> Example:
> ```shell
> # on packages/dart_frog, release dart frog v2 (lol)
> ../../tool/release_ready.sh 2.0.0 
> ```
> 
> This will generate a Changelog and create a release branch. Review the changelog, make the desired changes, commit, push, and open a PR. 
> After the PR is merged, create a [release on GitHub][github_release_link]. A Workflow should take care of publishing the new version on pub.dev.

Reviewer: What do you think of providing a step by step list?

```suggestion
Before starting the release process of an individual package, first check:

1. If your local `main` branch is up to date:
```shell
# ☁️ Ensure you're up to date with the GitHub remote
git checkout main
git fetch
git status
/```

2. Ensure the [GitHub pipeline](https://github.com/VeryGoodOpenSource/dart_frog/actions) is green (has passed successfuly) for your given package.

3. Checkout on `main` and run the script under `./tool/release_ready.sh` in the package root repository with the desired new version. 

```shell
# Run the release ready script (from packages/dart_frog)
../../tool/release_ready.sh 2.0.0 
/```

This will update the version of dart_frog to 2.0.0, update the dart_frog CHANGELOG.md, create and checkout to a local release branch.

4. Review the recently updated CHANGELOG file. You should manually amend the content were necessary. For example, by removing the redundant scope of some semantic pull requests or removing superfluous or unrelated logged changes.

5. Commit, push and a pull request from the new release branch.

6. Once merged, create a [release on GitHub][github_release_link]. A Workflow should take care of publishing the new version on the appropriate package manager.

7. Open follow-up pull requests updating this package usage in any other Dart Frog package that depends on this new release.
```

> Note, I've used /``` to terminate code blocks since I couldn't find a way to avoid terminating the suggestion block.

This is the internal package used by the Dart Frog tooling to interpret the file disposition and from it construct a Dart Frog server. 

> :warning: **Warning**: this package is a dependency on the bricks bundled into the CLI. This means that any changes that break the bricks should be released with a major version, otherwise dart frog users may be blocked from performing tasks such as `dev`, `build`, and `new`.

### `dart_frog_cli`

A Dart command line interface package that serves as the main tool for Dart Frog. It includes bundled versions of the bricks under `bricks/`. To sync the source code of the bricks with new bundles, run `tool/generate_bundles.sh`.

### Companion packages

The other items under `packages/` are companion packages in which dart_frog users may include on their project for specific server-side capabilities, such as auth (`dart_frog_auth`) and WebSockets (`dart_frog_web_socket`)

## Releasing versions of packages

Before starting the release process of an individual package, first check:

1. If your local `main` branch is up to date:
```shell
# ☁️ Ensure you're up to date with the GitHub remote
git checkout main
git fetch
git status
/```

> 2. Ensure the [GitHub pipeline](https://github.com/VeryGoodOpenSource/dart_frog/actions) is green (has passed successfuly) for your given package.

Reviewer: ```suggestion
2. Ensure the [GitHub pipeline](https://github.com/VeryGoodOpenSource/dart_frog/actions) is green (has passed successfully) for your given package.
```
